### PR TITLE
Improve spectral solver on staggered grids

### PR DESCRIPTION
**Pull request by EZoni**
**Goals**
- Implement high-order interpolation for momentum-conserving field gathering:
  - Source/Parallelization/WarpXComm.cpp
  - Source/Parallelization/WarpXComm_K.H
- Implement Galilean PSATD equations on staggered grids:
  - Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.H 
  - Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.cpp
- Implement comoving PSATD equations (only formulation with rho):
  - Source/FieldSolver/SpectralSolver/SpectralAlgorithms/ComovingPsatdAlgorithm.H
  - Source/FieldSolver/SpectralSolver/SpectralAlgorithms/ComovingPsatdAlgorithm.cpp

**To-do**
- [x] High-order interpolation using Fornberg coefficients
- [x] Galilean PSATD on staggered grids: update equations with `psatd.update_with_rho = 1`
- [x] Galilean PSATD on staggered grids: update equations with `psatd.update_with_rho = 0`
- [x] Galilean PSATD on staggered grids: current correction
- [x] Comoving PSATD: update equations with `psatd.update_with_rho = 1` (`psatd.update_with_rho = 0` in separate PR)
- [x] Comoving PSATD: current correction
- [x] Comoving PSATD: invert velocity sign
- [x] Clarify when to call `FillBoundaryAux`
- [x] Fix compilation bug in single precision
- [x] Update documentation

**Builds**
- [x] Compiles on CPUs
- [x] Compiles on GPUs

**Tests**
- [x] CI tests on CPUs
- [x] LWFA test in 2D on CPUs (with `psatd.update_with_rho = 1`)
- [x] LWFA test in 3D on GPUs (with `psatd.update_with_rho = 1`)
- [x] 2D regression test (LWFA, only default benchmark regression, no analysis) for staggered Galilean PSATD
- [x] 2D regression test (LWFA, only default benchmark regression, no analysis) for staggered comoving PSATD
- [ ] TBD: 3D regression tests (LWFA, only default benchmark regression, no analysis) might become too expensive

NCI mitigation with the comoving scheme seems more sensitive to the resolution than in the Galilean case (see top/bottom left corner of plot on the right). At the next higher resolution (not used in the CI test for practical reasons), the convergence with respect to NCI mitigation between the two schemes seem to improve significantly already.

Resolution 96 x 704 (CI test):
<p align="center">
    
    
</p>

Resolution  192 x 1408:
<p align="center">
    
    
</p>

### CODE REVIEW

File: Source/Evolve/WarpXEvolve.cpp
Comment (EZoni) [inline Source/Evolve/WarpXEvolve.cpp]: Note: adding this call to `FillBoundaryAux` causes the benchmark of the CI tests `momentum-conserving-gather` and `PlasmaAccelerationMR` to slightly change (see commits https://github.com/ECP-WarpX/WarpX/pull/1468/commits/e498760fb353503d5346ecfd746178ba26a06908 and https://github.com/ECP-WarpX/WarpX/pull/1468/commits/5d972a55225eacf5476dc17dabc31e442dfd482c, respectively).

File: Source/Evolve/WarpXEvolve.cpp
Comment (EZoni) [inline Source/Evolve/WarpXEvolve.cpp]: Note: this call to `FillBoundaryAux` before `UpdateAuxilaryData` does not seem intuitive. Would it be safe to remove it? Does it make a difference in specific scenarios, for example, in simulations with more than one refinement level?

File: Source/Diagnostics/FullDiagnostics.cpp
Comment (EZoni) [inline Source/Diagnostics/FullDiagnostics.cpp]: Note: this call to `FillBoundaryAux` before `UpdateAuxilaryData` does not seem intuitive. Would it be safe to remove it? Does it make a difference in specific scenarios, for example, in simulations with more than one refinement level?

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/ComovingPsatdAlgorithm.cpp
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralAlgorithms/ComovingPsatdAlgorithm.cpp:25]: Note: these are the infinite-order k vectors (argument `n_order = -1`).

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.cpp
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.cpp:30]: Note: these are the finite-order modified k vectors, always computed with the assumption of centered grids (argument `nodal = true`), for both nodal and staggered simulations.

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.cpp
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.cpp]: Note: the term `k dot v` here has to be computed always with the centered modified k vectors, in order to work correctly with both nodal and staggered grids.

File: Source/Parallelization/WarpXComm.cpp
Comment (EZoni) [inline Source/Parallelization/WarpXComm.cpp:136]: Note: I guess it is safe to leave this as it is, but it would be good to have a second opinion.

File: Source/WarpX.cpp
Comment (EZoni) [inline Source/WarpX.cpp:743]: Note: the rho-free formulation of the comoving PSATD scheme is not yet implemented and I do not plan to do so in this PR.

**Review by RemiLehe — APPROVED** [review]

File: Source/Diagnostics/FullDiagnostics.cpp
Comment (RemiLehe) [reply Source/Diagnostics/FullDiagnostics.cpp]: Hm, yes, it does seem that this might be superfluous.
Could you, in a separate PR, try to remove that line and run the automated tests (esp. checking the Checksum for the MR cases)?
Also, could you track, in the `git` history, when this line was added? Maybe this way we will find some corresponding explanation.

File: Source/Evolve/WarpXEvolve.cpp
Comment (RemiLehe) [reply Source/Evolve/WarpXEvolve.cpp]: Yes, same comment as above.

File: Source/FieldSolver/SpectralSolver/SpectralSolver.cpp
Comment (RemiLehe) [inline Source/FieldSolver/SpectralSolver/SpectralSolver.cpp:70]: Just to make the logic slightly more intuitive, could you rewrite this as:
```
if ( v_comoving[0] != 0. || v_comoving[1] != 0. || v_comoving[2] !=0 ) {
    ! Code that creates a `ComovingPstadAlgorithm`
} else if ( v_galilean[0] != 0. || v_galilean[1] != 0. || v_galilean[2] !=0 ) {
    ! Code that creates a `GalileanAlgorithm`
} else {
    ! Code that creates a regular `PsatdAlgorithm`
}
```

File: Docs/source/running_cpp/parameters.rst
Comment (RemiLehe) [inline Docs/source/running_cpp/parameters.rst:1160]: Could you add the documentation for `v_comoving` in the list of parameters?

File: Docs/source/running_cpp/parameters.rst
Comment (RemiLehe) [inline Docs/source/running_cpp/parameters.rst:1164]: It seems that this parameter is ignored when using FDTD. Could you confirm?
If that is the case, could you explicitly state this in the documentation?

File: Source/Parallelization/WarpXComm_K.H
Comment (RemiLehe) [inline Source/Parallelization/WarpXComm_K.H]: Throughout this file, could you elaborate a bit more on the comments:
```
// FDTD variant: uses 2nd order interpolation
```
```
// PSATD variant: uses high-order interpolation (chosen by user)
```

File: Regression/Checksum/benchmarks_json/PlasmaAccelerationMR.json
Comment (RemiLehe) [inline Regression/Checksum/benchmarks_json/PlasmaAccelerationMR.json]: Do you know why the results were changed in this FDTD simulation? Is that because of the new `FillBoundaryAux`?

File: Regression/Checksum/benchmarks_json/momentum-conserving-gather.json
Comment (RemiLehe) [inline Regression/Checksum/benchmarks_json/momentum-conserving-gather.json]: Same question as on the previous file.

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/ComovingPsatdAlgorithm.cpp
Comment (RemiLehe) [reply Source/FieldSolver/SpectralSolver/SpectralAlgorithms/ComovingPsatdAlgorithm.cpp:25]: Thanks for pointing this out.
Could you add this as a comment in the code itself?

File: Source/Parallelization/WarpXComm.cpp
Comment (RemiLehe) [inline Source/Parallelization/WarpXComm.cpp:159]: Could we add an `amrex::Assert` to check that `field_gathering_nox == 2` (same in `y` and `z`) when `finest_level>0`?

**Review by RemiLehe — CHANGES_REQUESTED** [review]
Thanks a lot for this PR!
The code was very clear. I have a few minor comments (see inline).

File: Source/Diagnostics/FullDiagnostics.cpp
Comment (EZoni) [reply Source/Diagnostics/FullDiagnostics.cpp]: Yes, I will try to incorporate these changes within a separate PR, so that it is easier to assess their effects. In the meantime, I think the calls to `FillBoundaryAux` before `UpdateAuxilaryData` were introduced [here](https://github.com/ECP-WarpX/WarpX/pull/504/files#diff-140169a8e394017bac0dd16e554a7fd7a9f4aa5adfb0991fe37f821b395398d2).

File: Regression/Checksum/benchmarks_json/PlasmaAccelerationMR.json
Comment (EZoni) [reply Regression/Checksum/benchmarks_json/PlasmaAccelerationMR.json]: Yes, in particular it is the call mentioned in [this comment](https://github.com/ECP-WarpX/WarpX/pull/1468#discussion_r523272221) above. All benchmarks updates will be dealt with separately in #1522 now.

File: Regression/Checksum/benchmarks_json/momentum-conserving-gather.json
Comment (EZoni) [reply Regression/Checksum/benchmarks_json/momentum-conserving-gather.json]: Yes, in particular it is the call mentioned in [this comment](https://github.com/ECP-WarpX/WarpX/pull/1468#discussion_r523272221) above. All benchmarks updates will be dealt with separately in #1522 now.

**Comment by EZoni** [comment]
@RemiLehe I have implemented the first changes that you requested. Thank you for all the suggestions. Once #1522 is merged, I will rebase this PR and mark it ready for review (second round).

**Comment by EZoni** [comment]
The new benchmarks introduced in this PR might be wrong due to a small bug in Regression/Checksum/checksum.py, introduced by myself in #1200 while working around a `yt` bug.

The error comes from
https://github.com/ECP-WarpX/WarpX/blob/bedec348b35ac28cf85cfe588e0ffd618fb2a1f9/Regression/Checksum/checksum.py#L76-L81

In particular, the line `Q = ds.index.grids[lev][field].v.squeeze()` sets `lev` as first index, while that index refers actually to the box index. This does not make a difference for tests with only one mesh level and one box. It is instead wrong for tests with many boxes.

Since the `yt` bug seems to occur only when the Galilean velocity is positive and only in 2D and RZ geometries, I will change the relevant Galilean tests in a separate PR and get rid of the above workaround, and then rebase and reset the benchmarks here (which correspond to tests with two boxes on two MPI ranks).

Will remove once again the "draft" tag from this PR once the issue is fixed.

**Update**
The bug fix is implemented in #1536, which must be merged first, before rebasing the PR here.

**Review by RemiLehe — APPROVED** [review]
Thanks a lot for all these changes! This will be very useful!

**Pull request merged**
